koerce/patterns.py

```diff
@@ -300,12 +296,8 @@ def __and__(self, other: Pattern) -> AllOf:
         -------
         New pattern that matches if both of the patterns match.
         """
-        if isinstance(self, AllOf) and isinstance(other, AllOf):
-            return AllOf(*self.patterns, *other.patterns)
-        elif isinstance(self, AllOf):
-            return AllOf(*self.patterns, other)
-        elif isinstance(other, AllOf):
-            return AllOf(self, *other.patterns)
```

Comment:
I dunno, maybe the original but with casts is better, the entire implementation is in a single place and there's no method dispatching 🤔